crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -1,125 +1,104 @@
+    ) -> anyhow::Result<u64> {
+        let iter = perpetual_db
+            .objects
+            .iter()
+            .skip_to_last()
+            .reverse()
+            .group_by(|item| item.0 .0);
+
         let mut total_keys_scanned = 0;
         let mut total_objects_scanned = 0;
-        let mut num_versions_for_key = 0;
-        let mut object_id: Option<ObjectID> = None;
-        let mut delete_range_start: Option<VersionNumber> = None;
-        let mut delete_range_end: Option<VersionNumber> = None;
         let mut num_pending_wb_ops = 0;
         let mut total_pruned: u64 = 0;
-        let mut wb = objects.batch();
-        for (key, _value) in iter {
-            total_keys_scanned += 1;
-            if let Some(obj_id) = object_id {
-                if obj_id != key.0 {
-                    total_objects_scanned += 1;
-                    if let (Some(start_seq_num), Some(end_seq_num)) =
-                        (delete_range_start, delete_range_end)
-                    {
-                        let start = ObjectKey(obj_id, start_seq_num);
-                        let end = ObjectKey(obj_id, end_seq_num);
-                        if let Ok(new_wb) = wb.delete_range(objects, &start, &end) {
-                            wb = new_wb;
-                            num_pending_wb_ops += 1;
-                        } else {
-                            error!("Failed to invoke delete_range on write batch while compacting objects");
-                            wb = objects.batch();
-                            num_pending_wb_ops = 0;
-                            break;
-                        }
-                        if num_pending_wb_ops >= MAX_OPS_IN_ONE_WRITE_BATCH {
-                            if wb.write().is_err() {
-                                error!("Failed to commit write batch while compacting objects");
-                                wb = objects.batch();
-                                num_pending_wb_ops = 0;
-                                break;
-                            } else {
-                                info!("Committed write batch while compacting objects, keys scanned = {:?}, objects scanned = {:?}", total_keys_scanned, total_objects_scanned);
-                            }
-                            wb = objects.batch();
-                            num_pending_wb_ops = 0;
-                        }
+
+        let mut wb = perpetual_db.objects.batch();
+
+        for (obj_id, group) in &iter {
+            let mut count = 0;
+            let mut delete_range = None;
+            let mut indirect_objects: HashMap<_, i64> = HashMap::new();
+
+            for (key, value) in group {
+                count += 1;
+                if count > num_versions_to_retain {
+                    delete_range = delete_range
+                        .map(|(_, end)| (key.1, end))
+                        .or_else(|| Some((key.1, SequenceNumber::from(key.1.value() + 1))));
```

Comment:
right side of range delete is not included, this is default stating point in case if there's only one element to delete
i.e. if you need to drop version 5, it will produce range [5,6)
before code structure was different and several extra variables were used, so right side of range delete was always a version of some existing element